pkg/apis/resource/validation/validation.go

```diff
@@ -729,3 +738,77 @@ func truncateIfTooLong(str string, maxLen int) string {
+	var allErrs field.ErrorList
+	allErrs = append(allErrs, validateDriverName(device.Driver, fldPath.Child("driver"))...)
+	allErrs = append(allErrs, validatePoolName(device.Pool, fldPath.Child("pool"))...)
+	allErrs = append(allErrs, validateDeviceName(device.Device, fldPath.Child("device"))...)
+	deviceID := structured.MakeDeviceID(device.Driver, device.Pool, device.Device)
+	if !allocatedDevices.Has(deviceID) {
+		allErrs = append(allErrs, field.Invalid(fldPath, deviceID, "must be an allocated device in the claim"))
+	}
+	allErrs = append(allErrs, metav1validation.ValidateConditions(device.Conditions, fldPath.Child("conditions"))...)
+	if len(device.Data.Raw) > 0 { // Data is an optional field.
+		allErrs = append(allErrs, validateRawExtension(device.Data, fldPath.Child("data"), false)...)
+	}
+	allErrs = append(allErrs, validateNetworkDeviceData(device.NetworkData, fldPath.Child("networkData"))...)
+	return allErrs
+}
+
+// validateRawExtension validates RawExtension as in https://github.com/kubernetes/kubernetes/pull/125549/
+func validateRawExtension(rawExtension runtime.RawExtension, fldPath *field.Path, stored bool) field.ErrorList {
+	var allErrs field.ErrorList
+	var v any
+	if len(rawExtension.Raw) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath, ""))
+	} else if !stored && len(rawExtension.Raw) > resource.OpaqueParametersMaxLength {
+		// Don't even bother with parsing when too large.
+		// Only applies on create. Existing parameters are grand-fathered in
+		// because the limit was introduced in 1.32. This also means that it
+		// can be changed in the future.
+		allErrs = append(allErrs, field.TooLong(fldPath, "" /* unused */, resource.OpaqueParametersMaxLength))
+	} else if err := json.Unmarshal(rawExtension.Raw, &v); err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, "<value omitted>", fmt.Sprintf("error parsing data as JSON: %v", err.Error())))
+	} else if v == nil {
+		allErrs = append(allErrs, field.Required(fldPath, ""))
+	} else if _, isObject := v.(map[string]any); !isObject {
+		allErrs = append(allErrs, field.Invalid(fldPath, "<value omitted>", "parameters must be a valid JSON object"))
+	}
+	return allErrs
+}
+
+const interfaceNameMaxLength int = 256
+const hardwareAddressMaxLength int = 128
+
+func validateNetworkDeviceData(networkDeviceData *resource.NetworkDeviceData, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if networkDeviceData == nil {
+		return allErrs
+	}
+
+	if len(networkDeviceData.InterfaceName) > interfaceNameMaxLength {
+		allErrs = append(allErrs, field.TooLong(fldPath.Child("interfaceName"), "" /* unused */, interfaceNameMaxLength))
+	}
+
+	if len(networkDeviceData.HardwareAddress) > hardwareAddressMaxLength {
+		allErrs = append(allErrs, field.TooLong(fldPath.Child("hardwareAddress"), "" /* unused */, hardwareAddressMaxLength))
+	}
+
+	allErrs = append(allErrs, validateSet(networkDeviceData.IPs, -1,
```

Comment:
this is a personal feeling maybe,  but I feel these helpers does not help, it is easier to do a loop and validate each element of the array than trying to understand what `validateSet` does with two closures

```
// IPs are in cidr format
for _, ip := range networkDeviceData.IPs {
  prefix, err:= netip.ParsePrefix(ip)
 if err != nil {
    allErrs = append(allErrs, "INVALID IP %v" err)
}
```

can we be more simplistic?

Reply:
I see, you need to handle the duplication, I wonder if this will handle it https://github.com/kubernetes/kubernetes/pull/128240/files#r1832831510 for us

Reply:
This would not as we need to clean the CIDRs, an IPv6 can be written in different format `2001:db8::1/64 == 2001:0db8::1/64 == 2001:0db8:0000:0000:0000:0000:0000:0001/64`

Reply:
but not  a blocker, just comments to myself

Reply:
understood, thanks for clarifying